### ISSUE TITLE
Bug: GPS date and time incorrectly sync RTC

### ISSUE THREAD

**Issue by adamgarbo** [adamgarbo/Cryologger_Iceberg_Tracking_Beacon#7]
An issue was discovered in the field that caused in the Adafruit Feather M0's RTC (SAMD21) to be synchronized with an incorrect date and time from the Adafruit Ultimate GPS FeatherWing (MTK3339).

Upon examining the produced NMEA messages, it is revealed that when GPS is initially powered on it begins from a default date of January 5th, 2080:

```
14:29:09.847 -> $GPGGA,235955.800,,,,,0,00,,,M,,M,,*7D
14:29:09.916 -> $GPRMC,235955.800,V,,,,,0.00,0.00,050180,,,N*44
14:29:10.839 -> $GPGGA,235956.800,,,,,0,00,,,M,,M,,*7E
14:29:10.910 -> $GPRMC,235956.800,V,,,,,0.00,0.00,050180,,,N*47
14:29:11.856 -> $GPGGA,235957.800,,,,,0,00,,,M,,M,,*7F
14:29:11.891 -> $GPRMC,235957.800,V,,,,,0.00,0.00,050180,,,N*46
14:29:12.856 -> $GPGGA,235958.800,,,,,0,00,,,M,,M,,*70
14:29:12.893 -> $GPRMC,235958.800,V,,,,,0.00,0.00,050180,,,N*49
14:29:13.857 -> $GPGGA,235959.800,,,,,0,00,,,M,,M,,*71
14:29:13.895 -> $GPRMC,235959.800,V,,,,,0.00,0.00,050180,,,N*48
14:29:14.870 -> $GPGGA,000000.800,,,,,0,00,,,M,,M,,*70
14:29:14.908 -> $GPRMC,000000.800,V,,,,,0.00,0.00,060180,,,N*4A
14:29:15.842 -> $GPGGA,000001.800,,,,,0,00,,,M,,M,,*71
14:29:15.915 -> $GPRMC,000001.800,V,,,,,0.00,0.00,060180,,,N*4B
14:29:16.872 -> $GPGGA,000002.800,,,,,0,00,,,M,,M,,*72
14:29:16.909 -> $GPRMC,000002.800,V,,,,,0.00,0.00,060180,,,N*48
14:29:17.866 -> $GPGGA,000003.800,,,,,0,00,,,M,,M,,*73
14:29:17.902 -> $GPRMC,000003.800,V,,,,,0.00,0.00,060180,,,N*49
14:29:18.849 -> $GPGGA,000004.800,,,,,0,00,,,M,,M,,*74
14:29:18.886 -> $GPRMC,000004.800,V,,,,,0.00,0.00,060180,,,N*4E
14:29:19.847 -> $GPGGA,000005.800,,,,,0,00,,,M,,M,,*75
```

An interesting limitation of the SAMD21’s RTC year register is that it is 5 bits, allowing for a maximum year value of 63. If a higher value is written to the register, it simply rolls over. This appears to explain why the date and time of Cryolgogers in the field were set to January 2016, but the question remains as to how and why the TinyGPS++ library deemed that a "fix" was acquired, despite not having valid data.

**Comment by rutvikbhimani** [comment]
Same issue found in SAMD21J18A
       vReadRTC >> MCU RTC : 05-01-2016 05:31:00
       vReadRTC >> GPS RTC :  05-01-2080 05:31:00

**Comment by adamgarbo** [comment]
Hi @rutvikbhimani,

Thank you for the input. I spoke with @mikalhart regarding this issue several months ago. He commented:

> In a $GPRMC sentence, the second ‘field’, variously called ‘[Navigation receiver warning](http://aprs.gids.nl/nmea/#rmc)’ or ‘[position status](https://docs.novatel.com/OEM7/Content/Logs/GPRMC.htm)’, can take one of two values: “A” for “Valid”, or “V” for “Invalid”.  Now there is some dispute whether “V” means “none of the data in this sentence is valid: ignore it all” or “ignore position information only: anything else is fair game”.  Through experimentation, TinyGPS++ has taken the latter view.  Because I’ve seen modules deliver a perfectly valid time and date for several minutes before changing that “V” over to “A”, I decided early on that I would report date and time regardless of that “validity” flag.

He goes on to suggest that we can test for this by changing the `GPS_SENTENCE_GPRMC` case in `TinyGPS++.cpp` from:

```
      case GPS_SENTENCE_GPRMC:
        date.commit();
        time.commit();
        if (sentenceHasFix)
        {
           location.commit();
           speed.commit();
           course.commit();
        }
        break;
```
to:
```
      case GPS_SENTENCE_GPRMC:
        if (sentenceHasFix)
        {
           date.commit();
           time.commit();
           location.commit();
           speed.commit();
           course.commit();
        }
        break;
```
 
This should force date.isUpdated() to return true only when the sentence has a good position fix.

However, even when I use a very restrictive if statement to try to weed out the bad dates and times, they still get through and are somehow still considered valid. This, in turn, causes the SAMD21 RTC year register to overflow when automatically synchronizing with the GPS.

```
if ((gps.location.isValid() && gps.date.isValid() && gps.time.isValid()) && 
    (gps.location.isUpdated() && gps.date.isUpdated() && gps.time.isUpdated())) { }
```

Cheers,
Adam

**Comment by adamgarbo** [comment]
The TinyGPS++ library has been inactive for a number of years and it is not anticipated that the issue(s) described here will be resolved by the author. A new approach to confirm the validity of messages received by the GNSS receiver has been developed that is based on custom GPGGA fix quality and GPRMC Navigation receiver warning objects. Tentatively closing this issue, as it should be addressed in the upcoming #9.
